app/slave/cluster_slave.py

```diff
@@ -56,6 +58,41 @@ def __init__(self, port, host, num_executors=10):
         self._build_teardown_coin = None
         self._base_executor_index = None
 
+        # Configure heartbeat
+        self._heartbeat_failure_count = 0
+        self._heartbeat_failure_threshold = Configuration['heartbeat_failure_threshold']
+        self._heartbeat_interval = Configuration['heartbeat_interval']
+        self._hb_scheduler = sched.scheduler()
+
+    def start_heartbeat_thread(self):
+        self._logger.info('Heartbeat will run every {} seconds'.format(self._heartbeat_interval))
+        SafeThread(target=self._start_heartbeat, name='HeartbeatThread', daemon=True).start()
```

Comment:
Sort've the same comment about SafeThread here, but i'm less confident that this is bad than the use of SafeThread in the master. We should probably talk about this.

1. Crashing a slave is much less devastating than crashing the master
2. We don't have alerts on the slave ... so maybe crashing the slave is the best we can do? Maybe we _shouldn't_ even make alerts on the slave. Maybe the proper response for a slave erring is just crashing the slave and relying on the master/kubernetes to spin up a new one. (hehe, kubernetes. I can dream)

Reply:
I would say it's _acceptable_ for this to be a SafeThread, but I wouldn't go so far as to say it _should_ be a SafeThread. Given that we're not assuming ephemeral slaves, the ideal for "maximum robustness" is that the master and slave should be able to tolerate arbitrarily long periods of time without being able to communicate with one another. When communication becomes possible again, they should seamlessly reconnect. What happens if there's a network partition between the master and slaves that lasts an hour? What happens if the master's vm dies and it takes us 30 minutes to bring it back up? I think the standard for perfection in those cases is that the slave seamlessly reconnects itself back to the master once the issue is resolved, not that the slave kills itself.

I say it's acceptable for the slave to kill itself is because in practice if the slave can't communicate with the master for more than ~5 minutes then something has probably gone seriously wrong and we probably will need an engineer to intervene anyway. But I do think intentionally killing the slave is intentionally giving up on perfection.

Reply:
> the master and slave should be able to tolerate arbitrarily long periods of time without being able to communicate with one another
> ...
> I think the standard for perfection in those cases is that the slave seamlessly reconnects itself back to the master once the issue is resolved

Let's discuss this offline. It sounds like we have very different thoughts on what the behavior should be in the case of network failure.